intercom-attributes/tests/data/com_interface.target.rs

```diff
@@ -193,55 +257,100 @@ impl Foo for ::intercom::ComItf<Foo>
+                    Err(::intercom::get_last_error(__result))
+                }
+            } )
+        } )();
+
+        #[allow(unused_imports)]
+        use ::intercom::ReturnError;
+        match result {
+            Ok( v ) => v,
+            Err( err ) => <ComResult<bool> as ReturnError>::handle(
+                ::intercom::return_hresult(err),
+            ),
         }
     }
 
     fn string_method(&self, msg: String) -> String
     {
+        #[allow(unused_imports)]
+        use ::intercom::ComInto;
+
         let comptr = ::intercom::ComItf::ptr(self);
         let vtbl = comptr as *const *const __FooVtbl;
 
+        let mut __msg_temporary =
+                <&::intercom::BStr as ::intercom::FromWithTemporary<String>>::to_temporary(msg)?;
+
         #[allow(unused_unsafe)]
-        unsafe {
-            let __result = ((**vtbl).string_method)( comptr, msg.into() );
-            __result.into()
+        let result : Result<String, ::intercom::ComError> = (|| unsafe {
+            let __result =
+                         ((**vtbl).string_method)(comptr,
+                                                  <&::intercom::BStr as ::intercom::FromWithTemporary<String>>::from_temporary(&mut __msg_temporary)?.as_ptr());
+            Ok( { ::intercom::BString::from_ptr( __result ).com_into()? } )
+        } )();
+
+        #[allow(unused_imports)]
+        use ::intercom::ReturnError;
+        match result {
+            Ok( v ) => v,
+            Err( err ) => <String as ReturnError>::handle(
+                ::intercom::return_hresult(err),
+            ),
         }
     }
 
     fn comitf_method(&self, itf: ComItf<Foo>) -> ComResult<ComItf<IUnknown>>
     {
+        #[allow(unused_imports)]
+        use ::intercom::ComInto;
+
         let comptr = ::intercom::ComItf::ptr(self);
         let vtbl = comptr as *const *const __FooVtbl;
 
         #[allow(unused_unsafe)]
-        unsafe {
+        let result : Result<ComResult<ComItf<IUnknown>>, ::intercom::ComError> = (|| unsafe {
             let mut __out: ComItf<IUnknown> = ComItf::null_itf();
             let __result = ((**vtbl).comitf_method)(comptr, itf.into(), &mut __out);
 
-            // Normal Result, not a ComResult. The Ok-case goes as before.
-            //
-            // The Err-case will use IErrorInfo to construct the original
-            // error type.
-            if __result == ::intercom::S_OK {
-                Ok(__out.into() )
-            } else {
-                Err(::intercom::get_last_error(__result))
-            }
+            Ok( {
+                // Normal Result, not a ComResult. The Ok-case goes as before.
+                //
+                // The Err-case will use IErrorInfo to construct the original
+                // error type.
+                if __result == ::intercom::S_OK {
+                    Ok(__out.into() )
+                } else {
+                    Err(::intercom::get_last_error(__result))
+                } 
```

Comment:
Duplication of the logic that I pointed out earlier. More convinced about the need of a helper function.

Reply:
Given Intercom generates a lot of code, a lot of the logic is "duplicated".

Also this is code that needs to be defined in the user code so any "helper function" it calls needs to be exposed publicly by Intercom. What kind of a signature did you have in mind? I guess we could give `HRESULT` a `map` function:

```rust
fn map<TOk, F: FnOnce() -> TOk>( self, f: F ) -> Result<TOk, ComError> {
    match self {
        S_OK => Ok( f() ),
        _ => Err( get_last_error( self ) )
    }
}
```

Reply:
I'll implement this in #74 on the `ComError` instead of doing it now for `HRESULT`.

Don't want to grow this PR any more.